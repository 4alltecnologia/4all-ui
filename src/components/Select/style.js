import styled, { createGlobalStyle } from 'styled-components';
import { default as theme } from '../../styles/variables';
import { flexPosition, fontStyle } from '../../styles/mixins';
import { lightenDarkenColor } from '../../helpers/helpers';

export const groupStyles = {
  display: 'flex',
  alignItems: 'center',
};

export const GroupLabel = {
  fontFamily: theme.fontFamily,
  fontSize: theme.fontSizes.SIZE_6,
  fontWeight: theme.fontWeights.SEMIBOLD,
  color: theme.colors.GRAY_MEDIUM,
  textTransform: 'initial',
};

export const customStyles = ({ error, customStyle = {}, canRemove }) => ({
  option: (stylesProvided, { isFocused, isDisabled }) => ({
    ...stylesProvided,
    fontFamily: theme.fontFamily,
    color: isDisabled ? theme.colors.GRAY_MEDIUM_3 : theme.colors.GRAY_DARKER,
    cursor: isDisabled ? 'not-allowed' : 'pointer',
    textDecoration: isDisabled ? 'line-through' : '',
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    backgroundColor: isFocused ? theme.colors.GRAY_LIGHT : '#FFF',
    height: customStyle.optionHeight || theme.input.DEFAULT_HEIGHT,
    ':active': {
      backgroundColor: theme.colors.GRAY_LIGHT_2,
    },
    label: {
      fontSize: theme.fontSizes.SIZE_6,
      fontWeight: 'normal',
    },
  }),
  control: () => ({
    width: '100%',
    height: customStyle.height || 40,
    display: 'flex',
    border: `1px solid ${error ? theme.colors.DANGER_COLOR : theme.colors.GRAY_LIGHT_2}`,
    borderRadius: theme.borders.RADIUS_1,
  }),
  multiValue: styles => ({
    ...styles,
    backgroundColor: customStyle.valueBgColor || theme.colors.MAIN_COLOR,
    borderRadius: '15px',
    padding: '0px 0px 0px 5px',
    fontSize: theme.fontSizes.SIZE_7,
    fontWeight: theme.fontWeights.SEMIBOLD,
    fontFamily: theme.fontFamily,
    marginRight: '3px',
    ...customStyle.customValueStyles,
  }),
  multiValueLabel: styles => ({
    ...styles,
    color: customStyle.valueColor || '#FFF',
    fontFamily: theme.fontFamily,
    fontSize: theme.fontSizes.SIZE_7,
    fontWeight: theme.fontWeights.SEMIBOLD,
    padding: 5,
    paddingRight: 0,
    paddingLeft: '2px',
  }),
  multiValueRemove: styles => !canRemove
    ? ({
      display: 'none',
    }) : ({
      ...styles,
      color: customStyle.multiRemoveBgColor || '#FFF',
      borderTopRightRadius: '15px',
      borderBottomRightRadius: '15px',
      cursor: 'pointer',
      paddingLeft: '2px',
      marginLeft: '2px',
      ':hover': {
        backgroundColor: lightenDarkenColor(
          customStyle.valueBgColor
          || theme.colors.MAIN_COLOR,
        ),
      },
    }),
  placeholder: styles => ({
    ...styles,
    fontSize: theme.fontSizes.SIZE_7,
    color: theme.colors.GRAY_MEDIUM,
    fontFamily: theme.fontFamily,
    fontWeight: theme.fontWeights.SEMIBOLD,
  }),
  valueContainer: styles => ({
    ...styles,
    paddingRight: '20px',
    /* flexWrap: 'nowrap', */
    '.select__single-value': {
      fontSize: theme.fontSizes.SIZE_7,
      fontWeight: theme.fontWeights.SEMIBOLD,
      color: customStyle.valueColor || theme.colors.GRAY_DARKER,
      backgroundColor: customStyle.valueBgColor || 'transparent',
      fontFamily: theme.fontFamily,
      ...customStyle.customValueStyles,
    }
  }),
});

export const QtdOptions = styled.span`
  ${flexPosition({ align: 'center', justify: 'center' })}
  ${fontStyle(theme.colors.GRAY_DARKER, theme.fontSizes.SIZE_7, theme.fontWeights.SEMIBOLD)}
  position: absolute;
  right: 0px;
  border-radius: 100%;
  padding: 0px 4px;
  height: 20px;
`;

export const GlobalSelectStyle = createGlobalStyle`
  .react-dropdown-multiselect {
    width: 100%;
  }
`;
