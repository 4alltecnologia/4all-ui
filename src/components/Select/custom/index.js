import React, { Fragment } from 'react';
import { components } from 'react-select';
import IconCheck from '../../Icons/Check';
import IconChevron from '../../Icons/Chevron';
import * as stylesSelect from '../style';

export const Option = (option, customStyle) => {
  return (
    <components.Option {...option}>
      <label>{option.label}</label>
      {option.isSelected ? <IconCheck color={customStyle.optionIconColor} /> : null}
    </components.Option>
  );
};

export const DropdownIndicator = (props, customStyle) => {
  return (
    components.DropdownIndicator && (
      <components.DropdownIndicator {...props}>
        <IconChevron color={customStyle.iconColor} />
      </components.DropdownIndicator>
    )
  );
};

export const MultiValueLabel = (props, formatLabel) => {
  /* eslint-disable */
  const options = props.selectProps.value;
  const currOption = props.data;
  return formatLabel ? (
    <components.MultiValueLabel id={`multi-value-${currOption.value}`} {...props}>
      {formatLabel(currOption, options)}
    </components.MultiValueLabel>
  ) : (
    <Fragment>
      <components.MultiValueLabel id={`multi-value-${currOption.value}`} {...props} />
    </Fragment>
  );
  /* eslint-enable */
};

export const formatGroupLabel = data => (
  <div style={stylesSelect.groupStyles}>
    <span style={stylesSelect.GroupLabel}>{data.label}</span>
  </div>
);

export const ValueContainer = (
  { children, ...props } // eslint-disable-line
) => (
  <components.ValueContainer {...props}>
    {children}
    <stylesSelect.QtdOptions className="multi-value-qtd-options" />
  </components.ValueContainer>
);
