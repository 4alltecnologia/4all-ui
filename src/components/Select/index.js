import React, { PureComponent, Fragment } from 'react';
import ReactSelect from 'react-select';
import PropTypes from 'prop-types';
import * as stylesSelect from './style';
import styles from '../../styles/variables';
import { formatSelectValues } from '../../helpers/select';
import {
  formatGroupLabel,
  DropdownIndicator,
  MultiValueLabel,
  ValueContainer,
  Option,
} from './custom';

class Select extends PureComponent {
  componentDidMount() {
    this.shouldFormatSelectValues();
  }

  componentDidUpdate() {
    this.shouldFormatSelectValues();
  }

  shouldFormatSelectValues = () => {
    const { isMultiple } = this.props;
    if (isMultiple) {
      formatSelectValues();
    }
  }

  render() {
    const {
      id,
      value,
      name,
      onChange,
      isMultiple,
      options,
      placeholder,
      formatLabel,
      error,
      closeMenuOnSelect,
      customValueStyles,
      canRemove,
      valueBgColor,
      valueColor,
      multiRemoveBgColor,
      iconColor,
      height,
      optionCustomDisabledStyles,
      optionHeight,
      containerCustomStyles,
    } = this.props;
    const style = {
      valueBgColor,
      valueColor,
      multiRemoveBgColor,
      customValueStyles,
      iconColor,
      height,
      optionHeight,
    };
    return (
      <Fragment>
        <stylesSelect.GlobalSelectStyle />
        <ReactSelect
          id={`select-${id}`}
          name={`${name}-${id}`}
          onChange={onChange}
          options={options}
          value={value}
          isMulti={isMultiple}
          styles={stylesSelect.customStyles({
            error,
            customStyle: style,
            canRemove,
            optionCustomDisabledStyles,
            containerCustomStyles,
          })}
          className="react-dropdown-multiselect"
          placeholder={placeholder}
          classNamePrefix="select"
          isClearable={false}
          clearable={false}
          hideSelectedOptions={false}
          closeMenuOnSelect={closeMenuOnSelect}
          formatGroupLabel={formatGroupLabel}
          isOptionDisabled={opt => opt.disabled}
          components={{
            Option: props => Option(props, style),
            DropdownIndicator: props => DropdownIndicator(props, style),
            IndicatorSeparator: () => null,
            ValueContainer,
            MultiValueLabel: props => MultiValueLabel(props, formatLabel),
          }}
        />
      </Fragment>
    );
  }
}

Select.propTypes = {
  id: PropTypes.string,
  value: PropTypes.any,
  name: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  error: PropTypes.bool,
  options: PropTypes.array,
  isMultiple: PropTypes.bool,
  placeholder: PropTypes.string,
  formatLabel: PropTypes.func,
  closeMenuOnSelect: PropTypes.bool,
  canRemove: PropTypes.bool,
  customValueStyles: PropTypes.object,
  valueBgColor: PropTypes.string,
  valueColor: PropTypes.string,
  multiRemoveBgColor: PropTypes.string,
  iconColor: PropTypes.string,
  height: PropTypes.string,
  optionHeight: PropTypes.string,
  optionCustomDisabledStyles: PropTypes.object,
  containerCustomStyles: PropTypes.object,
};

Select.defaultProps = {
  options: [],
  isMultiple: false,
  placeholder: 'Selecione',
  error: false,
  closeMenuOnSelect: false,
  formatLabel: null,
  id: 'dropdown',
  name: '',
  value: null,
  canRemove: true,
  customValueStyles: {},
  valueBgColor: null,
  valueColor: null,
  multiRemoveBgColor: null,
  iconColor: styles.colors.MAIN_COLOR,
  height: styles.input.DEFAULT_HEIGHT,
  optionHeight: styles.input.DEFAULT_HEIGHT,
  optionCustomDisabledStyles: {},
  containerCustomStyles: {},
};

export default Select;
