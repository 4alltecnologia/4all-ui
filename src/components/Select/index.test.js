import React from 'react';
import { shallow } from 'enzyme';
import Select from '.';

const options = [
  {
    id: 4864,
    label: 'Promoções',
    value: 4864,
  },
  {
    id: 4865,
    label: 'Sabores de Inverno',
    value: 4865,
  },
  {
    id: 4866,
    label: 'Combinados',
    value: 4866,
  },
];

describe('<Select />', () => {
  it('should render the component correctly', () => {
    const component = shallow(
      <Select id="test" name="test" options={options} onChange={() => {}} />,
    );

    expect(component).toMatchSnapshot();
  });

  it('should render the component with the properties', () => {
    const placeholder = 'Placeholder teste';
    const component = shallow(
      <Select
        id="test"
        name="select"
        options={options}
        placeholder={placeholder}
        onChange={() => console.log('ok')}
      />,
    );

    const stateManager = component.find('StateManager').props();

    expect(stateManager.options).toHaveLength(3);
    expect(stateManager.id).toEqual('select-test');
    expect(stateManager.name).toEqual('select-test');
    expect(stateManager.placeholder).toEqual(placeholder);
  });
});
