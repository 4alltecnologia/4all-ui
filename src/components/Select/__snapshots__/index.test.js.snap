// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should render the component correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Select
    canRemove={true}
    closeMenuOnSelect={false}
    customValueStyles={Object {}}
    error={false}
    formatLabel={null}
    height="40px"
    iconColor="#4FA444"
    id="test"
    isMultiple={false}
    multiRemoveBgColor={null}
    name="test"
    onChange={[Function]}
    optionHeight="40px"
    options={
        Array [
            Object {
              "id": 4864,
              "label": "Promoções",
              "value": 4864,
            },
            Object {
              "id": 4865,
              "label": "Sabores de Inverno",
              "value": 4865,
            },
            Object {
              "id": 4866,
              "label": "Combinados",
              "value": 4866,
            },
          ]
    }
    placeholder="Selecione"
    value={null}
    valueBgColor={null}
    valueColor={null}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <GlobalStyleComponent />,
        <StateManager
          className="react-dropdown-multiselect"
          classNamePrefix="select"
          clearable={false}
          closeMenuOnSelect={false}
          components={
                    Object {
                              "DropdownIndicator": [Function],
                              "IndicatorSeparator": [Function],
                              "MultiValueLabel": [Function],
                              "Option": [Function],
                              "ValueContainer": [Function],
                            }
          }
          defaultInputValue=""
          defaultMenuIsOpen={false}
          defaultValue={null}
          formatGroupLabel={[Function]}
          hideSelectedOptions={false}
          id="select-test"
          isClearable={false}
          isMulti={false}
          isOptionDisabled={[Function]}
          name="test-test"
          onChange={[Function]}
          options={
                    Array [
                              Object {
                                "id": 4864,
                                "label": "Promoções",
                                "value": 4864,
                              },
                              Object {
                                "id": 4865,
                                "label": "Sabores de Inverno",
                                "value": 4865,
                              },
                              Object {
                                "id": 4866,
                                "label": "Combinados",
                                "value": 4866,
                              },
                            ]
          }
          placeholder="Selecione"
          styles={
                    Object {
                              "control": [Function],
                              "multiValue": [Function],
                              "multiValueLabel": [Function],
                              "multiValueRemove": [Function],
                              "option": [Function],
                              "placeholder": [Function],
                              "valueContainer": [Function],
                            }
          }
          value={null}
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "react-dropdown-multiselect",
          "classNamePrefix": "select",
          "clearable": false,
          "closeMenuOnSelect": false,
          "components": Object {
            "DropdownIndicator": [Function],
            "IndicatorSeparator": [Function],
            "MultiValueLabel": [Function],
            "Option": [Function],
            "ValueContainer": [Function],
          },
          "defaultInputValue": "",
          "defaultMenuIsOpen": false,
          "defaultValue": null,
          "formatGroupLabel": [Function],
          "hideSelectedOptions": false,
          "id": "select-test",
          "isClearable": false,
          "isMulti": false,
          "isOptionDisabled": [Function],
          "name": "test-test",
          "onChange": [Function],
          "options": Array [
            Object {
              "id": 4864,
              "label": "Promoções",
              "value": 4864,
            },
            Object {
              "id": 4865,
              "label": "Sabores de Inverno",
              "value": 4865,
            },
            Object {
              "id": 4866,
              "label": "Combinados",
              "value": 4866,
            },
          ],
          "placeholder": "Selecione",
          "styles": Object {
            "control": [Function],
            "multiValue": [Function],
            "multiValueLabel": [Function],
            "multiValueRemove": [Function],
            "option": [Function],
            "placeholder": [Function],
            "valueContainer": [Function],
          },
          "value": null,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <GlobalStyleComponent />,
          <StateManager
            className="react-dropdown-multiselect"
            classNamePrefix="select"
            clearable={false}
            closeMenuOnSelect={false}
            components={
                        Object {
                                    "DropdownIndicator": [Function],
                                    "IndicatorSeparator": [Function],
                                    "MultiValueLabel": [Function],
                                    "Option": [Function],
                                    "ValueContainer": [Function],
                                  }
            }
            defaultInputValue=""
            defaultMenuIsOpen={false}
            defaultValue={null}
            formatGroupLabel={[Function]}
            hideSelectedOptions={false}
            id="select-test"
            isClearable={false}
            isMulti={false}
            isOptionDisabled={[Function]}
            name="test-test"
            onChange={[Function]}
            options={
                        Array [
                                    Object {
                                      "id": 4864,
                                      "label": "Promoções",
                                      "value": 4864,
                                    },
                                    Object {
                                      "id": 4865,
                                      "label": "Sabores de Inverno",
                                      "value": 4865,
                                    },
                                    Object {
                                      "id": 4866,
                                      "label": "Combinados",
                                      "value": 4866,
                                    },
                                  ]
            }
            placeholder="Selecione"
            styles={
                        Object {
                                    "control": [Function],
                                    "multiValue": [Function],
                                    "multiValueLabel": [Function],
                                    "multiValueRemove": [Function],
                                    "option": [Function],
                                    "placeholder": [Function],
                                    "valueContainer": [Function],
                                  }
            }
            value={null}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "react-dropdown-multiselect",
            "classNamePrefix": "select",
            "clearable": false,
            "closeMenuOnSelect": false,
            "components": Object {
              "DropdownIndicator": [Function],
              "IndicatorSeparator": [Function],
              "MultiValueLabel": [Function],
              "Option": [Function],
              "ValueContainer": [Function],
            },
            "defaultInputValue": "",
            "defaultMenuIsOpen": false,
            "defaultValue": null,
            "formatGroupLabel": [Function],
            "hideSelectedOptions": false,
            "id": "select-test",
            "isClearable": false,
            "isMulti": false,
            "isOptionDisabled": [Function],
            "name": "test-test",
            "onChange": [Function],
            "options": Array [
              Object {
                "id": 4864,
                "label": "Promoções",
                "value": 4864,
              },
              Object {
                "id": 4865,
                "label": "Sabores de Inverno",
                "value": 4865,
              },
              Object {
                "id": 4866,
                "label": "Combinados",
                "value": 4866,
              },
            ],
            "placeholder": "Selecione",
            "styles": Object {
              "control": [Function],
              "multiValue": [Function],
              "multiValueLabel": [Function],
              "multiValueRemove": [Function],
              "option": [Function],
              "placeholder": [Function],
              "valueContainer": [Function],
            },
            "value": null,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
